iOS/MSMapsModules/MSMapsModules/MSMapGeoJsonParser/MSMapGeoJsonParser.m

```diff
@@ -251,42 +313,135 @@ + (MSGeoposition * _Nullable)parseGeoposition:(NSArray *)coordinates
+    altitudeReferenceSystem:
+        (MSMapAltitudeReferenceSystem)altitudeReferenceSystem
+                      error:(NSError **)error {
+  if (pathArray.count == 0) {
+    *error = [MSMapGeoJsonParser
+        makeGeoJsonError:
+            [@"Geopath must contain an array of positions. Instead saw: "
+                stringByAppendingString:[MSMapGeoJsonParser
+                                            arrayToString:pathArray]]];
+    return nil;
+  }
+
   NSArray<MSGeoposition *> *path = [[NSArray alloc] init];
-  MSMapAltitudeReferenceSystem altitudeReferenceSystem =
-      MSMapAltitudeReferenceSystemSurface;
   for (id obj in pathArray) {
-    if (![obj isKindOfClass:[NSArray class]]) {
-      *error = [MSMapGeoJsonParser
-          makeGeoJsonError:
-              [@"coordinates of LineString must contain an array of "
-               @"positions. Instead saw: "
-                  stringByAppendingFormat:@"%@", (NSString *)obj]];
-      return nil;
-    }
-
     NSArray *latLong = (NSArray *)obj;
-    MSGeoposition *position = [self parseGeoposition:latLong error:error];
+    bool useAltitude = false;
+    if (altitudeReferenceSystem == MSMapAltitudeReferenceSystemEllipsoid) {
+      useAltitude = true;
+    }
+    MSGeoposition *position = [self parseGeoposition:latLong
+                                         useAltitude:useAltitude
+                                               error:error];
     if (*error != nil) {
       return nil;
     }
-    if (latLong.count > 2) {
-      altitudeReferenceSystem = MSMapAltitudeReferenceSystemEllipsoid;
-    }
     path = [path arrayByAddingObject:position];
   }
   return [[MSGeopath alloc] initWithPositions:path
                       altitudeReferenceSystem:altitudeReferenceSystem];
 }
 
++ (void)verifyPolygonRing:(NSArray *)path error:(NSError **)error {
+  if (path.count < 4) {
+    NSString *array = @"[";
+    for (int i = 0; i < path.count - 1; i++) {
+      array = [array
+          stringByAppendingString:[[MSMapGeoJsonParser
+                                      arrayToString:[path objectAtIndex:i]]
+                                      stringByAppendingString:@", "]];
+    }
+    array = [array
+        stringByAppendingString:[[MSMapGeoJsonParser
+                                    arrayToString:[path
+                                                      objectAtIndex:path.count -
+                                                                    1]]
+                                    stringByAppendingString:@"]"]];
+    *error = [MSMapGeoJsonParser
+        makeGeoJsonError:
+            [@"Polygon ring must have at least 4 positions, and the first and "
+             @"last position must be the same. Instead saw: "
+                stringByAppendingString:array]];
+    return;
+  }
+  NSArray *firstPosition = [path firstObject];
+  NSArray *lastPosition = [path lastObject];
+  if ((firstPosition.count < 3 && lastPosition.count < 3) ||
```

Comment:
What is this check doing? Is it checking that first and last have the same number of items?

Reply:
Yes - edited for clarity.